rust/connlib/tunnel/src/tests.rs

```diff
@@ -0,0 +1,359 @@
+        transition: <Self::Reference as ReferenceStateMachine>::Transition,
+    ) -> Self::SystemUnderTest {
+        match transition {
+            Transition::AddCidrResource(r) => {
+                state.client.add_resources(&[ResourceDescription::Cidr(r)]);
+            }
+            Transition::SendICMPPacketToResource { src, dst } => {
+                let _maybe_transmit = state
+                    .client
+                    .encapsulate(icmp_request_packet(src, dst), state.now);
+
+                // TODO: Handle transmit (send to relay / gateway)
+            }
+            Transition::Tick { millis } => {
+                state.now += Duration::from_millis(millis);
+                state.client.handle_timeout(state.now);
+                state.gateway.handle_timeout(state.now);
+            }
+        };
+
+        for (time, expected_event) in &ref_state.client_new_events {
+            assert_eq!(time, &state.now);
+            assert_eq!(expected_event, &state.client.poll_event().unwrap());
+        }
+
+        // TODO: Drain `poll_transmit` and execute it.
+
+        state
+    }
+}
+
+/// Several helper functions to make the reference state more readable.
+impl ReferenceState {
+    fn will_send_new_connection_intent(&self, resource: &ResourceId) -> bool {
+        let Some(last_intent) = self.last_connection_intents.get(resource) else {
+            return true;
+        };
+
+        self.now.duration_since(*last_intent) >= Duration::from_secs(2)
+    }
+}
+
+/// Implementation of our reference state machine.
+///
+/// The logic in here represents what we expect the [`ClientState`] & [`GatewayState`] to do.
+/// Care has to be taken that we don't implement things in a buggy way here.
+/// After all, if your test has bugs, it won't catch any in the actual implementation.
+impl ReferenceStateMachine for ReferenceState {
+    type State = Self;
+    type Transition = Transition;
+
+    fn init_state() -> proptest::prelude::BoxedStrategy<Self::State> {
+        (any::<[u8; 32]>(), any::<[u8; 32]>(), Just(Instant::now()))
+            .prop_filter(
+                "client and gateway priv key must be different",
+                |(c, g, _)| c != g,
+            )
+            .prop_map(|(client_priv_key, gateway_priv_key, now)| Self {
+                now,
+                client_priv_key,
+                gateway_priv_key,
+                client_resources: IpNetworkTable::new(),
+                client_new_events: Default::default(),
+                last_connection_intents: Default::default(),
+                ip4_resource_ips: Default::default(),
+                ip6_resource_ips: Default::default(),
+            })
+            .boxed()
+    }
+
+    /// Generate a set of transitions from the current state.
+    fn transitions(state: &Self::State) -> proptest::prelude::BoxedStrategy<Self::Transition> {
+        let add_resource = cidr_resource(8).prop_map(Transition::AddCidrResource);
```

Comment:
We limit the size of the subnet to a host mask of 8 here, i.e. /24 for IPv4 and /120 for IPv6. Without this, iterating all IPs just takes an insane amount of CPU power for not really any benefit in terms of discovering bugs.